src/main/java/org/openstreetmap/atlas/checks/validation/areas/WaterbodyAndIslandSizeCheck.java

```diff
@@ -0,0 +1,310 @@
+        {
+            final Relation relation = (Relation) object;
+            // Get Relation members that are an Area, have either an inner or outer member,
+            // and have yet to be examined
+            final Set<RelationMember> relationMembers = relation.members().stream()
+                    .filter(member -> member.getEntity() instanceof Area)
+                    .filter(member -> !this.isFlagged(member.getEntity().getIdentifier()))
+                    .filter(this::isInnerOrOuterMember).collect(Collectors.toSet());
+
+            if (!relationMembers.isEmpty())
+            {
+                final List<String> flagInstructions = new ArrayList<>();
+                final Set<RelationMember> invalidRelationMembers = new HashSet<>();
+
+                // Append member and given instruction for Relation members that fail the surface
+                // area size test.
+                this.getInvalidRelationMembers(relationMembers, invalidRelationMembers,
+                        flagInstructions);
+
+                if (!flagInstructions.isEmpty())
+                {
+                    // Create a single flag with each member entity and its given instruction
+                    return Optional.of(flagInstructions).map(instruction -> new CheckFlag(
+                            this.getTaskIdentifier(relation), invalidRelationMembers.stream()
+                                    .map(RelationMember::getEntity).collect(Collectors.toSet()),
+                            instruction));
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Helper function for filtering invalid Relation members. Each Relation member have either an
+     * inner or outer role, and the inner role must not contain the natural=rock tag.
+     *
+     * @param member
+     *            - RelationMember
+     * @return true if member is valid
+     */
+    private boolean isInnerOrOuterMember(final RelationMember member)
+    {
+        return member.getRole().equals(RelationTypeTag.MULTIPOLYGON_ROLE_OUTER) || (member.getRole()
+                .equals(RelationTypeTag.MULTIPOLYGON_ROLE_INNER)
+                && !Validators.isOfType(member.getEntity(), NaturalTag.class, NaturalTag.ROCK));
+    }
+
+    /**
+     * This function calculates the surface area of an individual Relation Member & generates an
+     * instruction of it fails the surface area test.
+     *
+     * @param relationMembers
+     *            - Set of valid RelationMembers
+     * @param invalidRelationMembers
+     *            - New Set of RelationMembers that fail surface area test
+     * @param instructions
+     *            - Flag Instructions for RelationMembers that fail surface area test
+     */
+    private void getInvalidRelationMembers(final Set<RelationMember> relationMembers,
+            final Set<RelationMember> invalidRelationMembers, final List<String> instructions)
```

Comment:
This method has side effects. It is updating `invalidRelationMembers` and `instructions` that are passed in as parameters.